### PR TITLE
Reorganize cmake build

### PR DESCRIPTION
**Pull request by jeanlucf22**
Fix use of PROGRESS
Split CMakeLists.txt file between main dir and src

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+if(DO_MPI)
+  set(LATTE LATTE_MPI_${PRECISION})
+else()
+  set(LATTE LATTE_${PRECISION})
+endif()
+
+add_subdirectory(src)
+
+#install library and executable
+install(TARGETS ${LATTE} DESTINATION ${CMAKE_INSTALL_BINDIR})
```
Comment (junghans) [inline CMakeLists.txt:79]: It is called `${LATTE}` here, but just `latte` everywhere else.

**Comment by junghans** [comment]
The other problem with creating a CMakeFiles.txt in the maindir is that it can overwrite the existing Makefile and hence I put it in a subfolder to start with.

**Comment by cnegre** [comment]
That's a good point

File: CMakeLists.txt (hunk 1 above)
Comment (jeanlucf22) [reply CMakeLists.txt:79]: latte is the .a (library file), ${LATTE} is the executable. I did not change that

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -182,13 +105,8 @@ endforeach()
...
-  set(LATTE LATTE_${PRECISION})
-endif()
 add_executable(${LATTE} ${CMAKE_CURRENT_SOURCE_DIR}/../src/latte.F90)
 target_link_libraries(${LATTE} latte ${PROGRESS_LIBRARIES})
-target_include_directories(${LATTE} PRIVATE ${MPI_Fortran_INCLUDE_PATH} ${PROGRESS_INCLUDE_DIRS} ${BML_INCLUDE_DIRS})
-target_compile_definitions(${LATTE} PRIVATE XSYEV ${PRECISION}PREC GPU${GPUOPT} DBCSR_${DBCSR_OPT} MPI_${DO_MPI} PROGRESS${PROGRESS})
-install(TARGETS ${LATTE} DESTINATION ${CMAKE_INSTALL_BINDIR})
+
+target_include_directories(latte PRIVATE ${MPI_Fortran_INCLUDE_PATH} ${PROGRESS_INCLUDE_DIRS} ${BML_INCLUDE_DIRS})
+target_compile_definitions(latte PRIVATE XSYEV ${PRECISION}PREC GPU${GPUOPT} DBCSR_${DBCSR_OPT} MPI_${DO_MPI} PROGRESS${PROGRESS})
```
Comment (junghans) [inline src/CMakeLists.txt:112]: @jeanlucf22 then something is wrong here. These defines and install rules were for the execuable.

**Comment by jeanlucf22** [comment]
This is the way most codes are organized, including PRORESS and BML. The way to run cmake it to run "cmake .." from within a build directory one level below the main directory. This way all the generated files are in that build directory, with all the temporary files. It is good practice to do it this way.

File: src/CMakeLists.txt (hunk 2 above)
Comment (jeanlucf22) [reply src/CMakeLists.txt:112]: Building the executable uses the .a and does not need the includes, I think. Is it failing for you?

Comment (junghans) [reply src/CMakeLists.txt:112]: I haven't checked, but back when at least the defines where using `latte.F90`. We could of course make the defines `PUBLIC` on the library, then all consumers would get them too.

**Comment by junghans** [comment]
> This is the way most codes are organized, including PRORESS and BML. The way to run cmake it to run "cmake .." from within a build directory one level below the main directory. This way all the generated files are in that build directory, with all the temporary files. It is good practice to do it this way.

I don't disagree, but if you have two build-systems, it is a bit risky as in-source build (`cmake .` in the maindir) will overwrite our custom Makefile. For projects with multiple build-systems the `cmake` subfolder seems to standard (see `lammps`, `tinker`, `py-onnx`, `glew`). Anyhow, the best would be to drop one of the two.

Also the "modern" way to use CMake is:
```
cmake -B builddir -S .
cmake --build builddir
```
so the whole `..` business is a bit obsolete anyhow.

**Comment by jeanlucf22** [comment]
> 

Some of the codes I work with (Sundials, SAMRAI, Trilinos) have a CMakeList.txt in the main directory... Maybe because they don't offer a simple Makefile build... So I guess neither one is standard. I don't mind either way

File: src/CMakeLists.txt (hunk 2 above)
Comment (jeanlucf22) [reply src/CMakeLists.txt:112]: I don't know the history of the cmake build in LATTE, but it certainly was not working for me before I made these changes.

Comment (junghans) [reply src/CMakeLists.txt:112]: It worked back in 2020 ;-)

**Pull request merged**